pennylane/devices/qubit_mixed/measure.py

```diff
@@ -49,163 +53,154 @@ def _reshape_state_as_matrix(state, num_wires):
+        readout_errors (List[Callable]): List of channels to apply to each wire being measured
         to simulate readout errors.
 
     Returns:
-        TensorLike: expectation value of observable wrt the state.
+        TensorLike: the result of the measurement
     """
-    probs = calculate_probability(measurementprocess, state, is_state_batched, readout_errors)
-    eigvals = math.asarray(measurementprocess.eigvals(), dtype="float64")
-    # In case of broadcasting, `probs` has two axes and these are a matrix-vector products
-    return math.dot(probs, eigvals)
-
-
-# pylint: disable=unused-argument
-def calculate_reduced_density_matrix(
-    measurementprocess: StateMeasurement,
-    state: TensorLike,
-    is_state_batched: bool = False,
-    readout_errors: list[Callable] = None,
-) -> TensorLike:
-    """Get the state or reduced density matrix.
+    for op in measurementprocess.diagonalizing_gates():
+        state = apply_operation(op, state, is_state_batched=is_state_batched)
 
-    Args:
-        measurementprocess (StateMeasurement): measurement to apply to the state.
-        state (TensorLike): state to apply the measurement to.
-        is_state_batched (bool): whether the state is batched or not.
-        readout_errors (List[Callable]): List of channels to apply to each wire being measured
-            to simulate readout errors. These are not applied on this type of measurement.
+    if readout_errors is not None and measurementprocess.wires is not None:
+        for err_channel_fn in readout_errors:
+            for w in measurementprocess.wires:
+                # Here, we assume err_channel_fn(w) returns a quantum operation/channel like qml.BitFlip(...)
+                error_op = err_channel_fn(w)
+                state = apply_operation(error_op, state, is_state_batched=is_state_batched)
 
-    Returns:
-        TensorLike: state or reduced density matrix.
-    """
-    wires = measurementprocess.wires
-    state_reshaped_as_matrix = _reshape_state_as_matrix(
-        state, _get_num_wires(state, is_state_batched)
-    )
-    if not wires:
-        return state_reshaped_as_matrix
+    num_wires = _get_num_wires(state, is_state_batched)
+    wires = Wires(range(num_wires))
+    flattened_state = _reshape_state_as_matrix(state, num_wires)
+    is_StateMP = isinstance(measurementprocess, StateMP)
+    is_DensityMatrixMP = isinstance(measurementprocess, DensityMatrixMP)
+    if is_StateMP and not is_DensityMatrixMP:
+        measurementprocess = DensityMatrixMP(wires)
+    res = measurementprocess.process_density_matrix(flattened_state, wires)
 
-    return math.reduce_dm(state_reshaped_as_matrix, wires)
+    return math.convert_like(res, state)
 
 
-def calculate_probability(
-    measurementprocess: StateMeasurement,
+def csr_dot_products_density_matrix(
+    measurementprocess: ExpectationMP,
     state: TensorLike,
     is_state_batched: bool = False,
     readout_errors: list[Callable] = None,
```

Comment:
No, but here is the very tricky part that we followed a design pattern (as in `qml.devices.qubit.measure`) that we pass method back the the main routine of `measure.py` to serve as `measure_functio`. This caused the trouble that even for those cases that do not require readout_errors arg it is still necessary to have these parameters. A better way is to absorb all these optional args into a huge option dictionary but since nothing new here to refactor I just leave them there.
Personally I feel here doing everything in separate measure func handlers is a little bit over-designed, but this is also helpful if we want to add more branches to this part in the future.